### PR TITLE
Implemented safetensors checkpoints save/load for Trainer

### LINKED ISSUE #22478: Implement safetensors checkpoint loading for Trainer
### Feature request

At the moment, Trainer loads models with `torch.load` method directly onto the cpu:

(`Trainer._load_from_checkpoint` method)
```python
                ... 
                # We load the model state dict on the CPU to avoid an OOM error.
                state_dict = torch.load(os.path.join(resume_from_checkpoint, WEIGHTS_NAME), map_location="cpu")
                # workaround for FSDP bug https://github.com/pytorch/pytorch/issues/82963
                # which takes *args instead of **kwargs
                load_result = model.load_state_dict(state_dict, False)
                # release memory
                del state_dict
                self._issue_warnings_after_load(load_result)
                ...
```

Loading on cpu with safetensors is a lot faster, so this method should (?) be preffered if safetensors library is installed.

### Motivation

I care to speed up checkpointing process since I use it to store the best model checkpoint every time the metric is improved (via callback).

### Your contribution

The change should be straightforward, but I am not sure if safetensors checkpointing should be defaulted to if safetensors are installed, or configured manually like in `PreTrainedModel.from_pretrained` method.

### PR DESCRIPTION
**Pull request by viktor-shcherb**
# What does this PR do?

Fixes #22478

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by ViktorooReps** [comment]
Should be fine now. @sgugger can you have a look?

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -2078,15 +2084,18 @@ def _load_from_checkpoint(self, resume_from_checkpoint, model=None):
         if model is None:
             model = self.model
 
-        if not os.path.isfile(os.path.join(resume_from_checkpoint, WEIGHTS_NAME)) and not os.path.isfile(
-            os.path.join(resume_from_checkpoint, WEIGHTS_INDEX_NAME)
-        ):
+        config_file = os.path.join(resume_from_checkpoint, CONFIG_NAME)
+        weights_file = os.path.join(resume_from_checkpoint, WEIGHTS_NAME)
+        weights_index_file = os.path.join(resume_from_checkpoint, WEIGHTS_INDEX_NAME)
+        safe_weights_file = os.path.join(resume_from_checkpoint, SAFE_WEIGHTS_NAME)
```
Comment (sgugger) [inline src/transformers/trainer.py:2090]: There could also be a sharded checkpoint (with `SAFE_WEIGHTS_INDEX_NAME`) if the model is big.

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -2078,15 +2084,18 @@ def _load_from_checkpoint(self, resume_from_checkpoint, model=None):
...
+
+        if not any(map(os.path.isfile, [weights_file, safe_weights_file, weights_index_file])):
```
Comment (sgugger) [inline src/transformers/trainer.py:2092]: A list comprehension would be faster than map here (it's 4 elements I know, but still)
```suggestion
        if not any([os.path.isfile(f) for f in [weights_file, safe_weights_file, weights_index_file]]):
```

File: src/transformers/training_args.py [hunk 3]
```diff
@@ -720,6 +725,16 @@ class TrainingArguments:
             )
         },
     )
+    save_safetensors: Optional[bool] = field(
+        default=None,
```
Comment (sgugger) [inline src/transformers/training_args.py:729]: I wouldn't pick an unset default, especially as we plan to phase in safetensors as a core dependency. It's going to be very surprising to users to have checkpoints suddenly saved in safetensors format. We will move forward (at some point in the future) to `save_pretrained` defaulting to safetensors, so I think this should be `False` for now, and we will switch to `True` at the same time as `save_pretrained`.

File: tests/trainer/test_trainer.py [hunk 4]
```diff
@@ -345,8 +352,11 @@ def get_regression_trainer(a=0, b=0, double_output=False, train_len=64, eval_len
 
 
 class TrainerIntegrationCommon:
-    def check_saved_checkpoints(self, output_dir, freq, total, is_pretrained=True):
-        file_list = [WEIGHTS_NAME, "training_args.bin", "optimizer.pt", "scheduler.pt", "trainer_state.json"]
+    def check_saved_checkpoints(self, output_dir, freq, total, is_pretrained=True, safe_weights=None):
```
Comment (sgugger) [inline tests/trainer/test_trainer.py:355]: Same here for the default

**Review by sgugger — COMMENTED** [review]
Thanks a lot for working on this and nice new tests! My main comment is on the default of the new training arguments, but otherwise LGTM!

File: src/transformers/trainer.py (hunk 1 above)
Comment (ViktorooReps) [reply src/transformers/trainer.py:2091]: Done. I have also added safe loading for sharded checkpoints (saving has already been implemented if I am not mistaking). Can you have a look? `load_sharded_checkpoint` at `modeling_utils.py`

File: src/transformers/training_args.py (hunk 3 above)
Comment (ViktorooReps) [reply src/transformers/training_args.py]: Done

File: tests/trainer/test_trainer.py (hunk 4 above)
Comment (ViktorooReps) [reply tests/trainer/test_trainer.py]: Done

**Comment by ViktorooReps** [comment]
One test failed:
```
=================================== FAILURES ===================================
_________________ WhisperModelTest.test_equivalence_pt_to_flax _________________
[gw1] linux -- Python 3.8.12 /home/circleci/.pyenv/versions/3.8.12/bin/python

self = <tests.models.whisper.test_modeling_whisper.WhisperModelTest testMethod=test_equivalence_pt_to_flax>

    @is_pt_flax_cross_test
    def test_equivalence_pt_to_flax(self):
        config, inputs_dict = self.model_tester.prepare_config_and_inputs_for_common()
        init_shape = (1,) + inputs_dict["input_features"].shape[1:]
    
        for model_class in self.all_model_classes:
            with self.subTest(model_class.__name__):
                fx_model_class_name = "Flax" + model_class.__name__
    
                if not hasattr(transformers, fx_model_class_name):
                    # no flax model exists for this class
                    return
    
                # Output all for aggressive testing
                config.output_hidden_states = True
                config.output_attentions = self.has_attentions
    
                fx_model_class = getattr(transformers, fx_model_class_name)
    
                # load PyTorch class
                pt_model = model_class(config).eval()
                # Flax models don't use the `use_cache` option and cache is not returned as a default.
                # So we disable `use_cache` here for PyTorch model.
                pt_model.config.use_cache = False
    
                # load Flax class
                fx_model = fx_model_class(config, input_shape=init_shape, dtype=jnp.float32)
    
                # make sure only flax inputs are forward that actually exist in function args
                fx_input_keys = inspect.signature(fx_model.__call__).parameters.keys()
    
                # prepare inputs
                pt_inputs = self._prepare_for_class(inputs_dict, model_class)
    
                # remove function args that don't exist in Flax
                pt_inputs = {k: v for k, v in pt_inputs.items() if k in fx_input_keys}
    
                # send pytorch inputs to the correct device
                pt_inputs = {
                    k: v.to(device=torch_device) if isinstance(v, torch.Tensor) else v for k, v in pt_inputs.items()
                }
    
                # convert inputs to Flax
                fx_inputs = {k: np.array(v) for k, v in pt_inputs.items() if torch.is_tensor(v)}
    
                fx_state = convert_pytorch_state_dict_to_flax(pt_model.state_dict(), fx_model)
                fx_model.params = fx_state
    
                # send pytorch model to the correct device
                pt_model.to(torch_device)
    
                with torch.no_grad():
                    pt_outputs = pt_model(**pt_inputs)
                fx_outputs = fx_model(**fx_inputs)
    
                fx_keys = tuple([k for k, v in fx_outputs.items() if v is not None])
                pt_keys = tuple([k for k, v in pt_outputs.items() if v is not None])
    
                self.assertEqual(fx_keys, pt_keys)
>               self.check_pt_flax_outputs(fx_outputs, pt_outputs, model_class)

tests/models/whisper/test_modeling_whisper.py:865: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tests/test_modeling_common.py:2098: in check_pt_flax_outputs
    self.check_pt_flax_outputs(
tests/test_modeling_common.py:2123: in check_pt_flax_outputs
    self.check_pt_flax_outputs(fx_output, pt_output, model_class, tol=tol, name=attr)
tests/test_modeling_common.py:2152: in check_pt_flax_outputs
    self.assertLessEqual(
E   AssertionError: 1.1086464e-05 not less than or equal to 1e-05 : outputs.encoder_last_hidden_state: Difference between PyTorch and Flax is 1.1086463928222656e-05 (>= 1e-05).
```
But I doubt that it is due to the changes added. @sgugger correct me if I am wrong

File: src/transformers/training_args.py [hunk 5]
```diff
@@ -261,6 +262,10 @@ class TrainingArguments:
         save_total_limit (`int`, *optional*):
             If a value is passed, will limit the total amount of checkpoints. Deletes the older checkpoints in
             `output_dir`.
+        save_safetensors (`bool`, *optional*):
```
Comment (sgugger) [inline src/transformers/training_args.py:265]:
```suggestion
        save_safetensors (`bool`, *optional*, defaults to `False`):
```

**Review by sgugger — APPROVED** [review]
Very nice! I have a couple more comments to finalize this PR, but it's very close to get merged!
And yes the test you mention is a flaky one.

File: src/transformers/modeling_utils.py [hunk 6]
```diff
@@ -358,10 +361,31 @@ def load_sharded_checkpoint(model, folder, strict=True):
...
     # Load the index
     index_file = os.path.join(folder, WEIGHTS_INDEX_NAME)
-    if not os.path.isfile(index_file):
-        raise ValueError(f"Can't find a checkpoint index ({WEIGHTS_INDEX_NAME}) in {folder}.")
+    safe_index_file = os.path.join(folder, SAFE_WEIGHTS_INDEX_NAME)
 
-    with open(index_file, "r", encoding="utf-8") as f:
+    index_present = os.path.isfile(index_file)
+    safe_index_present = os.path.isfile(safe_index_file)
+    safetensors_installed = is_safetensors_available()
```
[the remark this answers is not in the document]
Comment (ViktorooReps) [reply src/transformers/modeling_utils.py]: Done

File: src/transformers/modeling_utils.py [hunk 7]
```diff
@@ -336,7 +336,7 @@ def shard_checkpoint(
     return shards, index
 
 
-def load_sharded_checkpoint(model, folder, strict=True):
+def load_sharded_checkpoint(model, folder, strict=True, prefer_safe=False):
```
[the remark this answers is not in the document]
Comment (ViktorooReps) [reply src/transformers/modeling_utils.py]: Done

**Comment by ViktorooReps** [comment]
Should be good to merge now

**Comment by sgugger** [comment]
Thanks again for your contribution!

**Pull request merged**